ocaml/xapi/xapi_pif.ml

```diff
@@ -796,25 +797,41 @@ let rec unplug ~__context ~self =
   if Db.PIF.get_capabilities ~__context ~self |> List.mem "fcoe" then
     assert_fcoe_not_in_use ~__context self;
 
-  let tunnel = Db.PIF.get_tunnel_transport_PIF_of ~__context ~self in
+  let tunnel = pif_rec.API.pIF_tunnel_transport_PIF_of in
   if tunnel <> []
   then begin
     debug "PIF is tunnel transport PIF... also bringing down access PIF";
     let tunnel = List.hd tunnel in
     let access_PIF = Db.Tunnel.get_access_PIF ~__context ~self:tunnel in
     unplug ~__context ~self:access_PIF
   end;
-  if Db.PIF.get_sriov_logical_PIF_of ~__context ~self <> [] then begin
-    debug "PIF is network SRIOV logical PIF, also bringing down vlan on top of it";
-    unplug_vlan_on_sriov ~__context ~self
-  end;
-  let sriov = Db.PIF.get_sriov_physical_PIF_of ~__context ~self in
-  if sriov <> [] then begin
-    debug "PIF is network SRIOV physical PIF, also bringing down SRIOV logical PIF";
-    let pif = Db.Network_sriov.get_logical_PIF ~__context ~self:(List.hd sriov)in
-    unplug ~__context ~self:pif
+
+  (* Only exclusive PIF types can be put into following pattern match *)
+  begin match Xapi_pif_helpers.get_pif_topo ~__context ~pif_rec with
+    | Bond_master bond :: _ ->
+      List.iter (fun slave ->
+          if Db.PIF.get_sriov_physical_PIF_of ~__context ~self:slave <> [] then begin
+            debug "PIF is bond master, one of its slaves is a network SRIOV physical PIF, \
+                   also bringing down the slave as network SRIOV physical PIF";
+            unplug ~__context ~self:slave
+          end
+      ) (Db.Bond.get_slaves ~__context ~self:bond)
+    | Network_sriov_logical _ :: _ ->
+      debug "PIF is network SRIOV logical PIF, also bringing down vlan on top of it";
+      unplug_vlan_on_sriov ~__context ~self
+    | Physical pif_rec :: _ when pif_rec.API.pIF_sriov_physical_PIF_of <> [] ->
+      let sriov = List.hd pif_rec.API.pIF_sriov_physical_PIF_of in
```

Comment:
You can keep or drop `when pif_rec.API.pIF_sriov_physical_PIF_of <> []` and do
```ocaml
| Physical pif_rec :: _ ->
    List.iter (fun sriov ->
        debug "PIF is network SRIOV physical PIF, also bringing down SRIOV logical PIF";
        let pif = Db.Network_sriov.get_logical_PIF ~__context ~self:sriov in
        unplug ~__context ~self:pif)
    pif_rec.API.pIF_sriov_physical_PIF_of
```
Also move the comment below after `_ -> ` and before returning unit so that is clearer what it is referring to.

Reply:
It looks like the location of comment makes confusing.  I moved it above. Actually, this comment refers to the "SR-IOV Physical  PIF".  Meanwhile, it might be a bond slave. And if it is, it must not be brought down below.